File: packages/php-sdk/src/OpenNotify.php
<?php

declare(strict_types=1);

namespace OpenNotify;

use InvalidArgumentException;
use OpenNotify\Types\Channel;

/**
 * OpenNotify PHP SDK client.
 *
 * Send notifications through SMS, Telegram, Email, Push, and WhatsApp
 * using a single unified API.
 *
 * @example
 * $client = new OpenNotify('sk_your_api_key');
 *
 * $result = $client->send([
 *     'channel' => 'sms',
 *     'provider' => 'eskiz',
 *     'recipient' => '+998901234567',
 *     'message' => 'Hello from OpenNotify!',
 * ]);
 *
 * echo "Notification ID: " . $result['notification_id'];
 */
class OpenNotify
{
    private const DEFAULT_BASE_URL = 'https://api.opennotify.dev/api/v1';
    private const DEFAULT_TIMEOUT = 30.0;

    private HttpClient $http;

    /**
     * Create a new OpenNotify client.
     *
     * @param string $apiKey Your OpenNotify API key
     * @param string $baseUrl Base URL for the API (optional)
     * @param float $timeout Request timeout in seconds (optional)
     * @throws InvalidArgumentException If apiKey is empty
     */
    public function __construct(
        string $apiKey,
        string $baseUrl = self::DEFAULT_BASE_URL,
        float $timeout = self::DEFAULT_TIMEOUT,
    ) {
        if (empty($apiKey)) {
            throw new InvalidArgumentException('apiKey is required');
        }

        $this->http = new HttpClient($baseUrl, $apiKey, $timeout);
    }

    /**
     * Send a notification.
     *
     * @param array{
     *     channel: string,
     *     provider: string,
     *     recipient: string,
     *     message: string,
     *     subject?: string,
     *     metadata?: array<string, mixed>
     * } $options Notification options
     * @return array{notification_id: string} Result with notification ID
     * @throws OpenNotifyException If the request fails
     *
     * @example
     * // Send SMS
     * $result = $client->send([
     *     'channel' => 'sms',
     *     'provider' => 'eskiz',
     *     'recipient' => '+998901234567',
     *     'message' => 'Your code: 123456',
     * ]);
     *
     * // Send Email
     * $result = $client->send([
     *     'channel' => 'email',
     *     'provider' => 'sendgrid',
     *     'recipient' => 'user@example.com',
     *     'message' => 'Welcome!',
     *     'subject' => 'Welcome to our service',
     * ]);
     */
    public function send(array $options): array
    {
        $channel = $options['channel'] ?? throw new InvalidArgumentException('channel is required');
        $provider = $options['provider'] ?? throw new InvalidArgumentException('provider is required');
        $recipient = $options['recipient'] ?? throw new InvalidArgumentException('recipient is required');
        $message = $options['message'] ?? throw new InvalidArgumentException('message is required');

        $recipientData = $this->buildRecipient($channel, $recipient);

        $payload = ['text' => $message];
        if (isset($options['subject'])) {
            $payload['subject'] = $options['subject'];
        }
        if (isset($options['metadata'])) {
            $payload['metadata'] = $options['metadata'];
        }

        $data = $this->http->post('/notifications/send', [
            'provider' => strtoupper($provider),
            'recipient' => $recipientData,
            'payload' => $payload,
        ]);

        return [
            'notification_id' => $data['notificationId'],
        ];
    }

    /**
     * Get notification details by ID.
     *
     * @param string $id Notification ID
     * @return array{
     *     id: string,
     *     status: string,
     *     channel: string,
     *     provider: string,
     *     recipient: string,
     *     payload: array{text: string, subject?: string},
     *     created_at: string,
     *     sent_at?: string,
     *     delivered_at?: string
     * } Notification details
     * @throws OpenNotifyException If the notification is not found or request fails
     */
    public function getNotification(string $id): array
    {
        $data = $this->http->get("/notifications/{$id}");
        return $this->normalizeNotification($data);
    }

    /**
     * Get the delivery status of a notification.
     *
     * @param string $id Notification ID
     * @return array{status: string} Status response
     * @throws OpenNotifyException If the notification is not found or request fails
     */
    public function getStatus(string $id): array
    {
        $data = $this->http->get("/notifications/{$id}/status");
        return [
            'status' => strtolower($data['status']),
        ];
    }

    /**
     * List notifications with pagination.
     *
     * @param int $page Page number (1-indexed)
     * @param int $limit Items per page
     * @return array{
     *     notifications: array<array>,
     *     total: int,
     *     page: int,
     *     limit: int
     * } Paginated list of notifications
     * @throws OpenNotifyException If the request fails
     */
    public function listNotifications(int $page = 1, int $limit = 20): array
    {
        $data = $this->http->get('/notifications', [
            'page' => $page,
            'limit' => $limit,
        ]);

        $notifications = [];
        foreach ($data['notifications'] ?? [] as $notification) {
            $notifications[] = $this->normalizeNotification($notification);
        }

        return [
            'notifications' => $notifications,
            'total' => $data['total'] ?? 0,
            'page' => $data['page'] ?? $page,
            'limit' => $data['limit'] ?? $limit,
        ];
    }

    /**
     * Build recipient object based on channel type.
     *
     * @param string $channel Channel name
     * @param string $recipient Recipient identifier
     * @return array<string, string> Recipient object
     */
    private function buildRecipient(string $channel, string $recipient): array
    {
        $field = match (strtolower($channel)) {
            'sms', 'whatsapp' => 'phone',
            'email' => 'email',
            'telegram' => 'telegramChatId',
            'push' => 'deviceToken',
            default => 'phone',
        };

        return [$field => $recipient];
    }

    /**
     * Normalize notification response from API.
     *
     * @param array<string, mixed> $data Raw notification data
     * @return array<string, mixed> Normalized notification
     */
    private function normalizeNotification(array $data): array
    {
        $payload = [
            'text' => $data['payload']['text'] ?? '',
        ];
        if (!empty($data['payload']['subject'])) {
            $payload['subject'] = $data['payload']['subject'];
        }

        $result = [
            'id' => $data['id'],
            'status' => strtolower($data['status']),
            'channel' => strtolower($data['channel']),
            'provider' => strtolower($data['provider']),
            'recipient' => $data['recipient'] ?? '',
            'payload' => $payload,
            'created_at' => $data['createdAt'] ?? '',
        ];

        if (!empty($data['sentAt'])) {
            $result['sent_at'] = $data['sentAt'];
        }
        if (!empty($data['deliveredAt'])) {
            $result['delivered_at'] = $data['deliveredAt'];
        }

        return $result;
    }
}

File: packages/go-sdk/opennotify.go
// Package opennotify provides a Go client for the OpenNotify API.
//
// OpenNotify is a unified notification API for sending messages via
// SMS, Telegram, Email, Push, and WhatsApp channels.
//
// Example usage:
//
//	client, err := opennotify.New("sk_your_api_key")
//	if err != nil {
//	    log.Fatal(err)
//	}
//
//	result, err := client.Send(context.Background(), opennotify.SendOptions{
//	    Channel:   opennotify.ChannelSMS,
//	    Provider:  opennotify.ProviderEskiz,
//	    Recipient: "+998901234567",
//	    Message:   "Hello from OpenNotify!",
//	})
//	if err != nil {
//	    log.Fatal(err)
//	}
//
//	fmt.Println("Notification ID:", result.NotificationID)
package opennotify

import (
    "context"
    "encoding/json"
    "fmt"
    "strconv"
    "time"
)

// Client is the OpenNotify API client.
type Client struct {
    http *httpClient
}

// Options contains client configuration options.
type Options struct {
    // BaseURL is the API base URL (optional, defaults to https://api.opennotify.dev/api/v1).
    BaseURL string

    // Timeout is the HTTP request timeout (optional, defaults to 30 seconds).
    Timeout time.Duration
}

// New creates a new OpenNotify client.
//
// Example:
//
//	client, err := opennotify.New("sk_your_api_key")
func New(apiKey string, opts ...Options) (*Client, error) {
    if apiKey == "" {
        return nil, ErrAPIKeyRequired
    }

    var opt Options
    if len(opts) > 0 {
        opt = opts[0]
    }

    return &Client{
        http: newHTTPClient(opt.BaseURL, apiKey, opt.Timeout),
    }, nil
}

// Send sends a notification.
//
// Example:
//
//	// Send SMS
//	result, err := client.Send(ctx, opennotify.SendOptions{
//	    Channel:   opennotify.ChannelSMS,
//	    Provider:  opennotify.ProviderEskiz,
//	    Recipient: "+998901234567",
//	    Message:   "Your code is 1234",
//	})
//
//	// Send Telegram message
//	result, err := client.Send(ctx, opennotify.SendOptions{
//	    Channel:   opennotify.ChannelTelegram,
//	    Provider:  opennotify.ProviderTelegramBot,
//	    Recipient: "123456789",
//	    Message:   "Hello from OpenNotify!",
//	})
//
//	// Send Email
//	result, err := client.Send(ctx, opennotify.SendOptions{
//	    Channel:   opennotify.ChannelEmail,
//	    Provider:  opennotify.ProviderSendGrid,
//	    Recipient: "user@example.com",
//	    Subject:   "Welcome!",
//	    Message:   "Thank you for signing up.",
//	})
func (c *Client) Send(ctx context.Context, opts SendOptions) (*SendResult, error) {
    if opts.Channel == "" {
        return nil, ErrChannelRequired
    }
    if opts.Provider == "" {
        return nil, ErrProviderRequired
    }
    if opts.Recipient == "" {
        return nil, ErrRecipientRequired
    }
    if opts.Message == "" {
        return nil, ErrMessageRequired
    }

    req := sendRequest{
        Channel:   opts.Channel,
        Provider:  opts.Provider,
        Recipient: buildRecipient(opts.Channel, opts.Recipient),
        Payload: payloadRequest{
            Text:     opts.Message,
            Subject:  opts.Subject,
            Metadata: opts.Metadata,
        },
    }

    body, err := c.http.post(ctx, "/notifications/send", req)
    if err != nil {
        return nil, err
    }

    var result SendResult
    if err := json.Unmarshal(body, &result); err != nil {
        return nil, fmt.Errorf("failed to parse response: %w", err)
    }

    return &result, nil
}

// GetNotification retrieves a notification by ID.
//
// Example:
//
//	notification, err := client.GetNotification(ctx, "notif_123")
//	if err != nil {
//	    log.Fatal(err)
//	}
//	fmt.Println("Status:", notification.Status)
func (c *Client) GetNotification(ctx context.Context, id string) (*Notification, error) {
    if id == "" {
        return nil, ErrNotificationIDRequired
    }

    body, err := c.http.get(ctx, "/notifications/"+id, nil)
    if err != nil {
        return nil, err
    }

    var notification Notification
    if err := json.Unmarshal(body, &notification); err != nil {
        return nil, fmt.Errorf("failed to parse response: %w", err)
    }

    return &notification, nil
}

// GetStatus retrieves the delivery status of a notification.
//
// Example:
//
//	status, err := client.GetStatus(ctx, "notif_123")
//	if err != nil {
//	    log.Fatal(err)
//	}
//
//	if status.Status == opennotify.StatusDelivered {
//	    fmt.Println("Message delivered!")
//	}
func (c *Client) GetStatus(ctx context.Context, id string) (*StatusResponse, error) {
    if id == "" {
        return nil, ErrNotificationIDRequired
    }

    body, err := c.http.get(ctx, "/notifications/"+id+"/status", nil)
    if err != nil {
        return nil, err
    }

    var status StatusResponse
    if err := json.Unmarshal(body, &status); err != nil {
        return nil, fmt.Errorf("failed to parse response: %w", err)
    }

    return &status, nil
}

// ListNotifications retrieves a paginated list of notifications.
//
// Example:
//
//	// Get first page with default limit
//	list, err := client.ListNotifications(ctx, nil)
//
//	// Get specific page
//	list, err := client.ListNotifications(ctx, &opennotify.ListOptions{
//	    Page:  2,
//	    Limit: 50,
//	})
//
//	for _, n := range list.Notifications {
//	    fmt.Printf("%s: %s\n", n.ID, n.Status)
//	}
func (c *Client) ListNotifications(ctx context.Context, opts *ListOptions) (*NotificationList, error) {
    query := make(map[string]string)

    if opts != nil {
        if opts.Page > 0 {
            query["page"] = strconv.Itoa(opts.Page)
        }
        if opts.Limit > 0 {
            query["limit"] = strconv.Itoa(opts.Limit)
        }
    }

    body, err := c.http.get(ctx, "/notifications", query)
    if err != nil {
        return nil, err
    }

    var list NotificationList
    if err := json.Unmarshal(body, &list); err != nil {
        return nil, fmt.Errorf("failed to parse response: %w", err)
    }

    return &list, nil
}

// buildRecipient creates the recipient object based on channel type.
func buildRecipient(channel Channel, recipient string) recipientRequest {
    switch channel {
    case ChannelSMS, ChannelWhatsApp:
        return recipientRequest{Phone: recipient}
    case ChannelEmail:
        return recipientRequest{Email: recipient}
    case ChannelTelegram:
        return recipientRequest{TelegramChatID: recipient}
    case ChannelPush:
        return recipientRequest{DeviceToken: recipient}
    default:
        return recipientRequest{Phone: recipient}
    }
}
